File: _workspace/current/engineering/asset-pipeline/motion-bench/lib/kinematic-gate.mjs
import { readFileSync } from 'node:fs';

const RAD_TO_DEG = 180 / Math.PI;
const EPSILON = 1e-12;

export class KinematicGateError extends Error {
  constructor(code, message) {
    super(`${code}: ${message}`);
    this.name = 'KinematicGateError';
    this.code = code;
  }
}

function unit(quaternion) {
  if (!Array.isArray(quaternion) || quaternion.length !== 4) throw new KinematicGateError('KG_QUATERNION', 'a quaternion must have four components');
  const values = quaternion.map(Number);
  const magnitude = Math.hypot(...values);
  if (!Number.isFinite(magnitude) || magnitude <= EPSILON) throw new KinematicGateError('KG_QUATERNION', 'a quaternion must be finite and non-zero');
  return values.map(value => value / magnitude);
}

export function angularDistanceDegrees(left, right) {
  const a = unit(left), b = unit(right);
  const dot = Math.abs(a.reduce((total, value, index) => total + value * b[index], 0));
  return dot >= 1 - 1e-12 ? 0 : 2 * Math.acos(Math.max(0, Math.min(1, dot))) * RAD_TO_DEG;
}

export function measureQuaternionTrack(quaternions) {
  const frames = quaternions.map(unit);
  if (!frames.length) throw new KinematicGateError('KG_EMPTY_TRACK', 'at least one quaternion sample is required');
  const totals = frames.map(frame => frames.reduce((total, other) => total + angularDistanceDegrees(frame, other), 0));
  const medoidIndex = totals.reduce((best, total, index) => total < totals[best] ? index : best, 0);
  const peakDeg = Math.max(...frames.map(frame => angularDistanceDegrees(frame, frames[medoidIndex])));
  const stepDeg = frames.slice(1).reduce((maximum, frame, index) => Math.max(maximum, angularDistanceDegrees(frames[index], frame)), 0);
  return { peakDeg, stepDeg, medoidIndex };
}

export function validateBoundsJson(bounds) {
  const required = ['schemaVersion', 'sampleFps', 'referencePoseMethod', 'boundsFullCohort', 'boundsByExcludedSource'];
  const missing = required.filter(key => !(key in bounds));
  if (missing.length) throw new KinematicGateError('KG_BOUNDS_SCHEMA', `missing required bounds fields: ${missing.join(', ')}`);
  if (bounds.sampleFps !== 24 || bounds.referencePoseMethod !== 'angular-medoid-v1') throw new KinematicGateError('KG_BOUNDS_PROTOCOL', 'bounds do not use frozen 24-Hz angular-medoid-v1 protocol');
  return bounds;
}

function validateProvenance(provenance) {
  const required = ['assetId', 'clipName', 'action', 'actionClass', 'encoding', 'sourceGroup'];
  const missing = required.filter(key => !(key in provenance));
  if (missing.length) throw new KinematicGateError('KG_PROVENANCE', `missing provenance fields: ${missing.join(', ')}`);
  if (!['local-rest-relative-quaternion-deltas', 'absolute-local-rotation'].includes(provenance.encoding)) throw new KinematicGateError('KG_ENCODING', `unsupported encoding: ${provenance.encoding}`);
  if (!provenance.sourceGroup || (typeof provenance.sourceGroup !== 'object') || (!('repoRelativePath' in provenance.sourceGroup) && !('generator' in provenance.sourceGroup))) throw new KinematicGateError('KG_UNKNOWN_SOURCE', 'sourceGroup needs a bench path or authored generator');
  return provenance;
}

export function buildClipProvenanceOverlay(clip, assetId = 'unarmed-core') { return validateProvenance({ assetId, ...clip }); }
export function buildClipProvenanceCharacter(clip, assetId) { return validateProvenance({ assetId, ...clip }); }

const GLB_MAGIC = 0x46546c67;
const JSON_CHUNK = 0x4e4f534a;
const BIN_CHUNK = 0x004e4942;
const TYPE_COUNTS = { SCALAR: 1, VEC2: 2, VEC3: 3, VEC4: 4, MAT4: 16 };
const COMPONENT_SIZES = { 5121: 1, 5123: 2, 5125: 4, 5126: 4 };
const COMPONENT_READERS = {
  5121: (buffer, offset) => buffer.readUInt8(offset),
  5123: (buffer, offset) => buffer.readUInt16LE(offset),
  5125: (buffer, offset) => buffer.readUInt32LE(offset),
  5126: (buffer, offset) => buffer.readFloatLE(offset),
};

export function readGlb(path) {
  const buffer = readFileSync(path);
  if (buffer.length < 20 || buffer.readUInt32LE(0) !== GLB_MAGIC) throw new KinematicGateError('KG_GLB', `${path} is not a GLB`);
  const totalLength = buffer.readUInt32LE(8);
  if (totalLength !== buffer.length) throw new KinematicGateError('KG_GLB', `${path} has an invalid total length`);
  let offset = 12;
  let json = null;
  let bin = null;
  while (offset < totalLength) {
    if (offset + 8 > totalLength) throw new KinematicGateError('KG_GLB', `${path} has a truncated chunk header`);
    const length = buffer.readUInt32LE(offset);
    const type = buffer.readUInt32LE(offset + 4);
    const end = offset + 8 + length;
    if (end > totalLength) throw new KinematicGateError('KG_GLB', `${path} has a truncated chunk`);
    const chunk = buffer.subarray(offset + 8, end);
    if (type === JSON_CHUNK) {
      if (json !== null) throw new KinematicGateError('KG_GLB', `${path} contains multiple JSON chunks`);
      try { json = JSON.parse(chunk.toString('utf8')); } catch (error) { throw new KinematicGateError('KG_GLB', `${path} has invalid JSON: ${error.message}`); }
    } else if (type === BIN_CHUNK) {
      if (bin !== null) throw new KinematicGateError('KG_GLB', `${path} contains multiple BIN chunks`);
      bin = chunk;
    }
    offset = end;
  }
  if (offset !== totalLength || !json || !bin) throw new KinematicGateError('KG_GLB', `${path} requires exactly one JSON and BIN chunk`);
  return { json, bin };
}

export function readAccessor(glb, index) {
  const accessor = glb.json.accessors?.[index];
  if (!accessor || accessor.sparse) throw new KinematicGateError('KG_GLB_ACCESSOR', `unsupported accessor ${index}`);
  const width = TYPE_COUNTS[accessor.type];
  const size = COMPONENT_SIZES[accessor.componentType];
  const reader = COMPONENT_READERS[accessor.componentType];
  const view = glb.json.bufferViews?.[accessor.bufferView];
  if (!width || !size || !reader || !view) throw new KinematicGateError('KG_GLB_ACCESSOR', `unsupported accessor ${index}`);
  const base = (view.byteOffset ?? 0) + (accessor.byteOffset ?? 0);
  const stride = view.byteStride ?? width * size;
  if (stride < width * size || base < 0 || base + Math.max(0, accessor.count - 1) * stride + width * size > glb.bin.length) throw new KinematicGateError('KG_GLB_ACCESSOR', `out-of-range accessor ${index}`);
  return Array.from({ length: accessor.count }, (_, row) => Array.from({ length: width }, (_, column) => reader(glb.bin, base + row * stride + column * size)));
}

function sampleQuaternion(times, values, interpolation, at) {
  if (times.length !== values.length || !times.length) throw new KinematicGateError('KG_GLB_ANIMATION', 'rotation sampler has mismatched input/output samples');
  if (interpolation === 'CUBICSPLINE') throw new KinematicGateError('KG_GLB_ANIMATION', 'CUBICSPLINE rotation samplers are unsupported by frozen linear protocol');
  if (values.some(value => value.length !== 4)) throw new KinematicGateError('KG_GLB_ANIMATION', 'rotation sampler must contain VEC4 quaternions');
  if (at <= times[0]) return unit(values[0]);
  if (at >= times.at(-1)) return unit(values.at(-1));
  const upper = times.findIndex(time => time >= at);
  const lower = upper - 1;
  if (!(times[upper] > times[lower])) throw new KinematicGateError('KG_GLB_ANIMATION', 'rotation times must be strictly increasing');
  if (interpolation === 'STEP') return unit(values[lower]);
  if (interpolation !== undefined && interpolation !== 'LINEAR') throw new KinematicGateError('KG_GLB_ANIMATION', `unsupported rotation interpolation: ${interpolation}`);
  return slerp(values[lower], values[upper], (at - times[lower]) / (times[upper] - times[lower]));
}

function samplesAt24Hz(times) {
  if (times.some((time, index) => !Number.isFinite(time) || (index && time <= times[index - 1]))) throw new KinematicGateError('KG_GLB_ANIMATION', 'rotation times must be finite and strictly increasing');
  const start = times[0], end = times.at(-1);
  const samples = [start];
  for (let frame = 1; start + frame / 24 < end - 1e-9; frame += 1) samples.push(start + frame / 24);
  if (end > start) samples.push(end);
  return samples;
}

export function measureGlb(path) {
  const glb = readGlb(path);
  const rows = [];
  for (const animation of glb.json.animations ?? []) {
    for (const channel of animation.channels ?? []) {
      if (channel.target?.path !== 'rotation') continue;
      const sampler = animation.samplers?.[channel.sampler];
      const node = glb.json.nodes?.[channel.target.node];
      if (!sampler || !node?.name) throw new KinematicGateError('KG_GLB_ANIMATION', `animation ${animation.name ?? '<unnamed>'} has an invalid rotation channel`);
      const times = readAccessor(glb, sampler.input).map(row => row[0]);
      const values = readAccessor(glb, sampler.output);
      const quaternions = samplesAt24Hz(times).map(time => sampleQuaternion(times, values, sampler.interpolation ?? 'LINEAR', time));
      rows.push({ clipName: animation.name ?? `animation-${rows.length}`, bone: node.name, sampleFps: 24, ...measureQuaternionTrack(quaternions) });
    }
  }
  if (!rows.length) throw new KinematicGateError('KG_GLB_ANIMATION', `${path} has no rotation tracks`);
  return rows;
}

function slerp(left, right, ratio) {
  const a = unit(left); let b = unit(right);
  let dot = a.reduce((total, value, index) => total + value * b[index], 0);
  if (dot < 0) { b = b.map(value => -value); dot = -dot; }
  dot = Math.max(-1, Math.min(1, dot));
  if (dot > 0.9995) return unit(a.map((value, index) => (1 - ratio) * value + ratio * b[index]));
  const theta = Math.acos(dot), divisor = Math.sin(theta);
  return a.map((value, index) => (Math.sin((1 - ratio) * theta) * value + Math.sin(ratio * theta) * b[index]) / divisor);
}

export function redistributeStepShortestArc(quaternions, maxStepDegrees) {
  if (!(maxStepDegrees > 0)) throw new KinematicGateError('KG_REDISPATCH_WINDOW', 'maxStepDegrees must be positive');
  if (!quaternions.length) throw new KinematicGateError('KG_EMPTY_TRACK', 'at least one quaternion sample is required');
  const output = [unit(quaternions[0])];
  for (const current of quaternions.slice(1)) {
    const previous = output.at(-1);
    const steps = Math.max(1, Math.ceil(angularDistanceDegrees(previous, current) / maxStepDegrees));
    for (let index = 1; index <= steps; index += 1) output.push(slerp(previous, current, index / steps));
  }
  return output;
}

export function runConformanceVectors(pathOrPayload) {
  const payload = typeof pathOrPayload === 'string' ? JSON.parse(readFileSync(pathOrPayload, 'utf8')) : pathOrPayload;
  if (payload?.schemaVersion !== 1 || payload?.protocol !== 'angular-medoid-v1') throw new KinematicGateError('KG_VECTOR_SCHEMA', 'expected kinematic conformance vectors v1');
  const tolerance = Number(payload.angleToleranceDeg ?? 0.1);
  const results = (payload.vectors ?? []).map(vector => {
    if (!vector.id || vector.times?.length !== vector.quaternions?.length) throw new KinematicGateError('KG_VECTOR_SCHEMA', `invalid vector: ${vector.id ?? '<unnamed>'}`);
    const actual = measureQuaternionTrack(vector.quaternions);
    const expected = vector.expected;
    if (actual.medoidIndex !== expected.medoidIndex || ['peakDeg', 'stepDeg'].some(key => Math.abs(actual[key] - expected[key]) > tolerance)) throw new KinematicGateError('KG_CONFORMANCE', `${vector.id} expected ${JSON.stringify(expected)}, got ${JSON.stringify(actual)}`);
    return { id: vector.id, actual };
  });
  if (results.map(result => result.id).join(',') !== 'V1,V2,V3,V4,V5') throw new KinematicGateError('KG_VECTOR_SCHEMA', 'Stage A requires exactly V1–V5');
  return results;
}

File: _workspace/current/engineering/asset-pipeline/tools/kinematic_gate.py
#!/usr/bin/env python3
"""Policy-free frozen quaternion measurement primitives for the motion bench."""
from __future__ import annotations

import json
import math
import struct
from pathlib import Path
from typing import Any, Iterable, Sequence

DEGREES = 180.0 / math.pi
EPSILON = 1e-12


class KinematicGateError(ValueError):
    """A fail-closed, machine-readable gate error."""

    def __init__(self, code: str, message: str):
        self.code = code
        super().__init__(f"{code}: {message}")


def _unit(quaternion: Sequence[float]) -> tuple[float, float, float, float]:
    if len(quaternion) != 4:
        raise KinematicGateError("KG_QUATERNION", "a quaternion must have four components")
    values = tuple(float(component) for component in quaternion)
    magnitude = math.sqrt(sum(component * component for component in values))
    if not math.isfinite(magnitude) or magnitude <= EPSILON:
        raise KinematicGateError("KG_QUATERNION", "a quaternion must be finite and non-zero")
    return tuple(component / magnitude for component in values)  # type: ignore[return-value]


def angular_distance_degrees(left: Sequence[float], right: Sequence[float]) -> float:
    """Return d(q1,q2)=2*acos(clamp(abs(dot(q1,q2)),0,1)) in degrees."""
    q_left, q_right = _unit(left), _unit(right)
    dot = abs(sum(a * b for a, b in zip(q_left, q_right)))
    return 0.0 if dot >= 1.0 - 1e-12 else 2.0 * math.acos(max(0.0, min(1.0, dot))) * DEGREES


def measure_quaternion_track(quaternions: Iterable[Sequence[float]]) -> dict[str, float | int]:
    """Measure angular-medoid peak and adjacent-sample step; ties choose earliest."""
    frames = [_unit(frame) for frame in quaternions]
    if not frames:
        raise KinematicGateError("KG_EMPTY_TRACK", "at least one quaternion sample is required")
    totals = [sum(angular_distance_degrees(frame, other) for other in frames) for frame in frames]
    medoid_index = min(range(len(frames)), key=lambda index: (totals[index], index))
    reference = frames[medoid_index]
    peak = max(angular_distance_degrees(frame, reference) for frame in frames)
    step = max((angular_distance_degrees(frames[index - 1], frames[index]) for index in range(1, len(frames))), default=0.0)
    return {"peakDeg": peak, "stepDeg": step, "medoidIndex": medoid_index}


def validate_bounds_json(bounds: dict[str, Any]) -> dict[str, Any]:
    """Reject incomplete stage-B bounds before a gate can select from them."""
    required = ("schemaVersion", "sampleFps", "referencePoseMethod", "boundsFullCohort", "boundsByExcludedSource")
    missing = [key for key in required if key not in bounds]
    if missing:
        raise KinematicGateError("KG_BOUNDS_SCHEMA", f"missing required bounds fields: {', '.join(missing)}")
    if bounds["sampleFps"] != 24 or bounds["referencePoseMethod"] != "angular-medoid-v1":
        raise KinematicGateError("KG_BOUNDS_PROTOCOL", "bounds do not use frozen 24-Hz angular-medoid-v1 protocol")
    return bounds


def _validate_provenance(provenance: dict[str, Any]) -> dict[str, Any]:
    required = ("assetId", "clipName", "action", "actionClass", "encoding", "sourceGroup")
    missing = [key for key in required if key not in provenance]
    if missing:
        raise KinematicGateError("KG_PROVENANCE", f"missing provenance fields: {', '.join(missing)}")
    if provenance["encoding"] not in {"local-rest-relative-quaternion-deltas", "absolute-local-rotation"}:
        raise KinematicGateError("KG_ENCODING", f"unsupported encoding: {provenance['encoding']!r}")
    source_group = provenance["sourceGroup"]
    if not isinstance(source_group, dict):
        raise KinematicGateError("KG_UNKNOWN_SOURCE", "sourceGroup must be an object")
    if "repoRelativePath" not in source_group and "generator" not in source_group:
        raise KinematicGateError("KG_UNKNOWN_SOURCE", "sourceGroup needs a bench path or authored generator")
    return provenance


def build_clip_provenance_overlay(clip: dict[str, Any], asset_id: str = "unarmed-core") -> dict[str, Any]:
    return _validate_provenance({"assetId": asset_id, **clip})


def build_clip_provenance_character(clip: dict[str, Any], asset_id: str) -> dict[str, Any]:
    return _validate_provenance({"assetId": asset_id, **clip})


_GLB_MAGIC = 0x46546C67
_JSON_CHUNK = 0x4E4F534A
_BIN_CHUNK = 0x004E4942
_TYPE_COUNTS = {"SCALAR": 1, "VEC2": 2, "VEC3": 3, "VEC4": 4, "MAT4": 16}
_COMPONENT_FORMATS = {5121: ("B", 1), 5123: ("H", 2), 5125: ("I", 4), 5126: ("f", 4)}


def read_glb(path: str | Path) -> dict[str, Any]:
    """Parse a single-buffer GLB strictly enough for the frozen motion metric."""
    raw = Path(path).read_bytes()
    if len(raw) < 20 or struct.unpack_from("<I", raw, 0)[0] != _GLB_MAGIC:
        raise KinematicGateError("KG_GLB", f"{path} is not a GLB")
    total_length = struct.unpack_from("<I", raw, 8)[0]
    if total_length != len(raw):
        raise KinematicGateError("KG_GLB", f"{path} has an invalid total length")
    offset, document, binary = 12, None, None
    while offset < total_length:
        if offset + 8 > total_length:
            raise KinematicGateError("KG_GLB", f"{path} has a truncated chunk header")
        length, chunk_type = struct.unpack_from("<II", raw, offset)
        end = offset + 8 + length
        if end > total_length:
            raise KinematicGateError("KG_GLB", f"{path} has a truncated chunk")
        chunk = raw[offset + 8:end]
        if chunk_type == _JSON_CHUNK:
            if document is not None:
                raise KinematicGateError("KG_GLB", f"{path} contains multiple JSON chunks")
            try:
                document = json.loads(chunk.decode("utf-8"))
            except (UnicodeDecodeError, json.JSONDecodeError) as error:
                raise KinematicGateError("KG_GLB", f"{path} has invalid JSON: {error}") from error
        elif chunk_type == _BIN_CHUNK:
            if binary is not None:
                raise KinematicGateError("KG_GLB", f"{path} contains multiple BIN chunks")
            binary = chunk
        offset = end
    if offset != total_length or document is None or binary is None:
        raise KinematicGateError("KG_GLB", f"{path} requires exactly one JSON and BIN chunk")
    return {"json": document, "bin": binary}


def read_accessor(glb: dict[str, Any], index: int) -> list[list[float]]:
    """Read an interleaved GLB accessor; sparse/data-URI accessors fail closed."""
    try:
        accessor = glb["json"]["accessors"][index]
        view = glb["json"]["bufferViews"][accessor["bufferView"]]
        component_format, component_size = _COMPONENT_FORMATS[accessor["componentType"]]
        width = _TYPE_COUNTS[accessor["type"]]
    except (IndexError, KeyError, TypeError) as error:
        raise KinematicGateError("KG_GLB_ACCESSOR", f"unsupported accessor {index}") from error
    if accessor.get("sparse"):
        raise KinematicGateError("KG_GLB_ACCESSOR", f"unsupported sparse accessor {index}")
    base = int(view.get("byteOffset", 0)) + int(accessor.get("byteOffset", 0))
    stride = int(view.get("byteStride", component_size * width))
    count = int(accessor["count"])
    if stride < component_size * width or base < 0 or base + max(0, count - 1) * stride + component_size * width > len(glb["bin"]):
        raise KinematicGateError("KG_GLB_ACCESSOR", f"out-of-range accessor {index}")
    return [
        [float(struct.unpack_from("<" + component_format, glb["bin"], base + row * stride + column * component_size)[0]) for column in range(width)]
        for row in range(count)
    ]


def _sample_quaternion(times: Sequence[float], values: Sequence[Sequence[float]], interpolation: str, moment: float) -> tuple[float, float, float, float]:
    if not times or len(times) != len(values) or any(len(value) != 4 for value in values):
        raise KinematicGateError("KG_GLB_ANIMATION", "rotation sampler has mismatched VEC4 samples")
    if interpolation == "CUBICSPLINE":
        raise KinematicGateError("KG_GLB_ANIMATION", "CUBICSPLINE rotation samplers are unsupported by frozen linear protocol")
    if moment <= times[0]:
        return _unit(values[0])
    if moment >= times[-1]:
        return _unit(values[-1])
    upper = next((index for index, value in enumerate(times) if value >= moment), None)
    if upper is None or upper == 0 or times[upper] <= times[upper - 1]:
        raise KinematicGateError("KG_GLB_ANIMATION", "rotation times must be strictly increasing")
    if interpolation == "STEP":
        return _unit(values[upper - 1])
    if interpolation != "LINEAR":
        raise KinematicGateError("KG_GLB_ANIMATION", f"unsupported rotation interpolation: {interpolation}")
    return _slerp(values[upper - 1], values[upper], (moment - times[upper - 1]) / (times[upper] - times[upper - 1]))


def _samples_at_24hz(times: Sequence[float]) -> list[float]:
    if not times or any(not math.isfinite(value) or (index and value <= times[index - 1]) for index, value in enumerate(times)):
        raise KinematicGateError("KG_GLB_ANIMATION", "rotation times must be finite and strictly increasing")
    output = [times[0]]
    frame = 1
    while times[0] + frame / 24.0 < times[-1] - 1e-9:
        output.append(times[0] + frame / 24.0)
        frame += 1
    if times[-1] > times[0]:
        output.append(times[-1])
    return output


def measure_glb(path: str | Path) -> list[dict[str, Any]]:
    """Measure real GLB rotation tracks at endpoint-preserving 24 Hz without bpy."""
    glb = read_glb(path)
    rows = []
    for animation in glb["json"].get("animations", []):
        for channel in animation.get("channels", []):
            target = channel.get("target", {})
            if target.get("path") != "rotation":
                continue
            try:
                sampler = animation["samplers"][channel["sampler"]]
                node = glb["json"]["nodes"][target["node"]]
                name = node["name"]
            except (IndexError, KeyError, TypeError) as error:
                raise KinematicGateError("KG_GLB_ANIMATION", f"invalid rotation channel in {animation.get('name', '<unnamed>')}") from error
            times = [row[0] for row in read_accessor(glb, sampler["input"])]
            values = read_accessor(glb, sampler["output"])
            quaternions = [_sample_quaternion(times, values, sampler.get("interpolation", "LINEAR"), moment) for moment in _samples_at_24hz(times)]
            rows.append({
                "clipName": animation.get("name", f"animation-{len(rows)}"), "bone": name, "sampleFps": 24,
                **measure_quaternion_track(quaternions),
            })
    if not rows:
        raise KinematicGateError("KG_GLB_ANIMATION", f"{path} has no rotation tracks")
    return rows


def _slerp(left: Sequence[float], right: Sequence[float], ratio: float) -> tuple[float, float, float, float]:
    a, b = _unit(left), _unit(right)
    dot = sum(x * y for x, y in zip(a, b))
    if dot < 0.0:
        b, dot = tuple(-value for value in b), -dot
    dot = max(-1.0, min(1.0, dot))
    if dot > 0.9995:
        return _unit(tuple((1.0 - ratio) * x + ratio * y for x, y in zip(a, b)))
    theta = math.acos(dot)
    divisor = math.sin(theta)
    return tuple((math.sin((1.0 - ratio) * theta) * x + math.sin(ratio * theta) * y) / divisor for x, y in zip(a, b))  # type: ignore[return-value]


def redistribute_step_shortest_arc(quaternions: Iterable[Sequence[float]], max_step_degrees: float) -> list[tuple[float, float, float, float]]:
    if max_step_degrees <= 0.0:
        raise KinematicGateError("KG_REDISPATCH_WINDOW", "max_step_degrees must be positive")
    source = list(quaternions)
    if not source:
        raise KinematicGateError("KG_EMPTY_TRACK", "at least one quaternion sample is required")
    output = [_unit(source[0])]
    for current in source[1:]:
        previous = output[-1]
        steps = max(1, math.ceil(angular_distance_degrees(previous, current) / max_step_degrees))
        output.extend(_slerp(previous, current, index / steps) for index in range(1, steps + 1))
    return output


def run_conformance_vectors(path: str | Path) -> list[dict[str, Any]]:
    payload = json.loads(Path(path).read_text())
    if payload.get("schemaVersion") != 1 or payload.get("protocol") != "angular-medoid-v1":
        raise KinematicGateError("KG_VECTOR_SCHEMA", "expected kinematic conformance vectors v1")
    tolerance = float(payload.get("angleToleranceDeg", 0.1))
    results = []
    for vector in payload.get("vectors", []):
        if not vector.get("id") or len(vector.get("times", [])) != len(vector.get("quaternions", [])):
            raise KinematicGateError("KG_VECTOR_SCHEMA", f"invalid vector: {vector.get('id', '<unnamed>')}")
        actual = measure_quaternion_track(vector["quaternions"])
        expected = vector["expected"]
        if actual["medoidIndex"] != expected["medoidIndex"] or any(abs(float(actual[key]) - float(expected[key])) > tolerance for key in ("peakDeg", "stepDeg")):
            raise KinematicGateError("KG_CONFORMANCE", f"{vector['id']} expected {expected}, got {actual}")
        results.append({"id": vector["id"], "actual": actual})
    ids = [result["id"] for result in results]
    if ids != ["V1", "V2", "V3", "V4", "V5"]:
        raise KinematicGateError("KG_VECTOR_SCHEMA", "Stage A requires exactly V1–V5")
    return results
